Task 2 - Line counter to json

thiantonello opened the pull request:
Add ruby gem json v.2.6.1

In parser.rb
    - Create count_lines method
    - Create count_lines_json method to generate a json with the file name and the number of lines
       
In parser_spec.rb
    - Create tests to both new methods

Modify main.rb to print count_lines_json

thiantonello committed:
build: add ruby gem json v.2.6.1

thiantonello committed:
feat: create new methods and tests to Parser class

In parser.rb
- Create count_lines method
- Create count_lines_json method to generate a json with the file name and the number of lines

In parser_spec.rb
- Create tests to both new methods

thiantonello committed:
feat: modify main.rb to print count_lines_json

mauricio-prl commented on lib/parser.rb:
```diff
@@ -9,4 +11,19 @@ def first_line
+
+  def count_lines_json
+    file = @file.split("/").last
+    lines = count_lines()
```
Nao precisa dos parenteses aqui

mauricio-prl commented on lib/parser.rb:
```diff
@@ -9,4 +11,19 @@ def first_line
+  end
+
+  def count_lines_json
+    file = @file.split("/").last
```
Ta meio confuso aqui, se vc recebe um path de um file, chama a sua variavel de instancia de `@path` ou `@file_path`. Dessa forma deu a entender que o seu `@file` ja era uma instancia de `File`

mauricio-prl commented on spec/lib/parser_spec.rb:
```diff
@@ -19,4 +19,30 @@
+      expect(subject.count_lines).to eq(158)
+    end
+
+    #verificar com os avaliadores/reviwers qual modo usar para testar o count_lines (de cima ou de baixo)
```
Se for seguir nessa abordagem eu faria o de cima, pra manter a descrição e o seu `subject` ser sempre a instancia e nao a execução do metodo. Mas da pra fazer da segunda forma tbm, eu faria assim:

```rb
describe "#count_lines" do
  subject { described_class.new(file_path).count_lines }

  let(:file_path) { "./spec/fixtures/games_test.log" }

  it { is_expected.to eq(158) }
end
```

https://relishapp.com/rspec/rspec-core/docs/subject/one-liner-syntax

thiantonello replied:
O teste do método count_lines foi removido porque ele se tornou um método privado.  De toda forma, agradeço a sugestão!

mauricio-prl commented on spec/lib/parser_spec.rb:
```diff
@@ -19,4 +19,30 @@
+  describe "#count_lines_json" do
+    subject { described_class.new("./spec/fixtures/games_test.log") }
+
+    let(:obj) { { "games_test.log" => {
```
se precisa quebrar a let em varias linhas usa bloco `do/end`

mauricio-prl commented on spec/lib/parser_spec.rb:
```diff
@@ -19,4 +19,30 @@
+    
+    specify { expect(JSON.parse(subject.count_lines_json)).to eq(obj) }
+    
+    #verificar com os avaliadores/reviewers qual modo usar para testar o count_lines_json (de cima ou de baixo)
```
eu prefiro a segunda, é mais comum

brunohkbx replied:
Pra mim não faz sentido nenhum especificar o subject dentro de um it inline, sendo que a intenção do it inline é de o utilizar com is_expected.

O correto seria:

`it { is_expected.to eq(JSON.pretty_generate(obj)) }`

brunohkbx commented on lib/parser.rb:
```diff
@@ -9,4 +11,19 @@ def first_line
     content
   end
+
+  def count_lines
```
Se esse método não for usado em outro lugar, deixe o privado

thiantonello committed:
feat: make count_lines private method and refactor

- count_lines is now a private method
- refactor the name of instance variable from @file to @path
to enhance code understanding and avoid confusion

thiantonello committed:
test: exclude private method test and refactor

mauricio-prl commented on lib/parser.rb:
```diff
@@ -1,12 +1,31 @@
     data.close
     content
   end
+
```
```suggestion
```

mauricio-prl commented on lib/parser.rb:
```diff
@@ -1,12 +1,31 @@
+      :lines => lines
+      } }
+      
+      JSON.pretty_generate(obj)
```
```suggestion
    JSON.pretty_generate(obj)
```

mauricio-prl commented on lib/parser.rb:
```diff
@@ -1,12 +1,31 @@
+      JSON.pretty_generate(obj)
+    end
+
+    private
```
```suggestion
    private
    
```

brunohkbx commented on lib/parser.rb:
```diff
@@ -1,12 +1,31 @@
+    lines = count_lines
+    
+    obj = { file => {
+      :lines => lines
```
```suggestion
      :lines => count_lines()
```

Se não me engano há uma regra no rubocop pra invocar funções usando o `()` pois poderia ficar ambíguo como se tivesse chamando por uma variável

brunohkbx commented on lib/parser.rb:
```diff
@@ -1,12 +1,31 @@
   end
+
+  
+  def count_lines_json
```
Não faz sentido chamar essa função de count_lines sendo que na próxima tarefa vc vai precisar expor mais informações dentro do json

thiantonello committed:
refactor: count_lines_json to generate_json
